src/components/scrollManager.js

```diff
@@ -0,0 +1,501 @@
+        if (element) {
+            var parent = element.parentElement;
+
+            while (parent) {
+                if ((!vertical && parent.scrollWidth > parent.clientWidth && parent.classList.contains("scrollX")) ||
+                    (vertical && parent.scrollHeight > parent.clientHeight) && parent.classList.contains("scrollY")) {
+                    return parent;
+                }
+
+                parent = parent.parentElement;
+            }
+        }
+
+        return documentScroller;
+    }
+
+    /**
+     * @typedef {Object} ScrollerData
+     * @property {number} scrollPos current scroll position
+     * @property {number} scrollSize scroll size
+     * @property {number} clientSize client size
+     */
+
+    /**
+     * Returns scroll data for specified orientation.
+     *
+     * @param {HTMLElement} scroller scroller
+     * @param {boolean} vertical vertical scroll data
+     * @return {ScrollerData} scroll data
+     */
+    function getScrollerData(scroller, vertical) {
+        var data = {};
+
+        if (!vertical) {
+            data.scrollPos = scroller.scrollLeft;
+            data.scrollSize = scroller.scrollWidth;
+            data.clientSize = scroller.clientWidth;
+        } else {
+            data.scrollPos = scroller.scrollTop;
+            data.scrollSize = scroller.scrollHeight;
+            data.clientSize = scroller.clientHeight;
+        }
+
+        return data;
+    }
+
+    /**
+     * Returns position of child of scroller for specified orientation.
+     *
+     * @param {HTMLElement} scroller scroller
+     * @param {HTMLElement} element child of scroller
+     * @param {boolean} vertical vertical scroll
+     * @return {number} child position
+     */
+    function getScrollerChildPos(scroller, element, vertical) {
+        var elementRect = element.getBoundingClientRect();
+        var scrollerRect = scroller.getBoundingClientRect();
+
+        if (!vertical) {
+            return scroller.scrollLeft + elementRect.left - scrollerRect.left;
+        } else {
+            return scroller.scrollTop + elementRect.top - scrollerRect.top;
+        }
+    }
+
+    /**
+     * Returns scroll position for element.
+     *
+     * @param {ScrollerData} scrollerData scroller data
+     * @param {number} elementPos child element position
+     * @param {number} elementSize child element size
+     * @param {boolean} centered scroll to center
+     * @return {number} scroll position
+     */
+    function calcScroll(scrollerData, elementPos, elementSize, centered) {
+        var maxScroll = scrollerData.scrollSize - scrollerData.clientSize;
+
+        var scroll;
+
+        if (centered) {
+            scroll = elementPos + (elementSize - scrollerData.clientSize) / 2;
+        } else {
+            var delta = fitRange(elementPos, elementPos + elementSize - 1, scrollerData.scrollPos, scrollerData.scrollPos + scrollerData.clientSize - 1);
+            scroll = scrollerData.scrollPos - delta;
```

Comment:
why is it so instead of `scroll = elementPos + delta`?

Reply:
`fitRange` returns delta to move `element` into `scroller` viewport. But, in fact, we should move `scroller` - move its viewport over `element`.